rfcs/001 - EC2 schema.md

```diff
@@ -0,0 +1,136 @@
+
+- Key Pairs. They can be assigned to an instance on creation time, so another FK column from the `instance` table to the key pair.
+
+- Network interfaces. Logical network components. An ENI belongs to an instance, but an instance can have multiple ENIs (similar to EBS). So the FK in this case will be on the ENI's side.
+
+- Auto Scaling Groups can be created manually, using launch templates (and launch configurations that are like launch templates just for ASG?) or can be created by other entities like EKS. Like spot requests, ASGs should not have a direct relation with `instance`s and can be part of the `instance_metadata`. To be able to know if an instance was created by an ASG we check the tags and look for the id to link them.
+
+- AMI's management: This can be complex, not only in performance and how we have experienced it before, but also an AMI can be owned by yourself or someone else, you can check for yours or third parties. They can be private or public. I've been thinking that maybe they should be their module but if we implement this it makes sense to stay in the EC2 module since the instance directly relates to the AMI in use.
+
+- Saving plans. Budget solution. Is not related only to EC2, but to Fargate and Lambda too. Commitment to a consistent amount of usage (measured in $/hour) for a 1 or 3-year term. The flow is similar to a marketplace. Maybe need its module? I do not foresee direct relation with any entity of the aws_ec2 module. In the AWS Console UI, it appears under EC2 but seems to be part of the AWS Cost Management service.
+
+![Saving plans](./assets/001/saving_plan.png)
+
+- Scheduled instances. Another marketplace-like service. The desired schedule needs to be set and then based on that you can see the offering, put it in your cart and then buy them.
+
+- Reserved instances. also a budget solution with a supply and demand-like model. You insert the requirement you want and then you can see the offering. Based on that offering you can pick and buy. It is like parking spaces, you have reserved one instance type and then when you create one that matches the specifications will be taken into account for the cost.
+
+After all these features, the EC2 schema should look something like the following:
+
+![Proposed schema](./assets/001/proposed_schema.png)
+
+Note: This schema does not have all the possible relationships, but the most important ones with the `instance` table.
+Note 2: Could be also accessed via https://dbdocs.io/alejandro/proposed-iasql?view=relationships
+
+Some of the features exposed do apply changes automatically to the cloud, like spot instances and Auto Scaling Groups. These will imply the development of the "safer sync faster apply" feature as a pre-requisite.
+
+we can start with instance type, launch templates, dedicated hosts, capacity reservation, ebs snapshots, placement groups, key pairs and network interfaces.
+
```

Comment:
I have one comment about instanceType. The types of instances are defined, and right now it is just an string. We could create a table and only allow to insert the defined values there.

Reply:
I definitely think that's a good idea.

It could also be done via the `aws_ec2_metadata` module to attach that safety onto the ec2 module. So it's still a string, but we attach a foreign key relationship to an `instance_type` table if the metadata module is installed so users can opt-in on the improved type safety.

Reply:
Yes, and also instance types depend on availability zones. This is explained in one of the first points of this proposal (https://github.com/iasql/iasql-engine/pull/1003/files#diff-35d35a757122d48b71d944384346f61d9c234056b348aa48f7072db57cdbd715R70).

This will be a table where the source of truth will always be the cloud. We have handled these cases by just deleting new records if a user creates one, and on updates/deletes just restore the records from the cloud.